File: Python/907. 子数组的最小值之和.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        __MOD__=10**9+7
        stack=[]
        left=[]
        for i in range(len(arr)):
            while stack and arr[i]<=arr[stack[-1]]:
                stack.pop()
            
            if not stack:
                left.append(-1)
            else:
                left.append(stack[-1])
            stack.append(i)
        
        stack.clear()
        right=[0]*len(arr)
        for i in range(len(arr)-1,-1,-1):
            while stack and arr[i]<arr[stack[-1]]:
                stack.pop()
            
            if not stack:
                right[i]=(len(arr))
            else:
                right[i]=(stack[-1])
            stack.append(i)
        #print(left)
        #print(right)
        ans=0
        for i in range(len(arr)):
            ans=(ans+(i-left[i])*(right[i]-i)*arr[i])%__MOD__
        return ans

File: JavaScript/907. 子数组的最小值之和.js
/**
 * @param {number[]} arr
 * @return {number}
 */
 var sumSubarrayMins = function(arr) {
    var ans=0;
    var __MOD__=10**9+7;

    function solve(l,r){
        if(l>r)return 0;
        if(l==r)return arr[l];
        var mi=arr[l];
        var midx=l;
        for(var i=l;i<=r;i++){
            if(arr[i]<mi){
                mi=arr[i];
                midx=i;
            }
        }
        var cur=(midx-l+1)*(r-midx+1)*mi;
        var ansl=solve(l,midx-1)%__MOD__;
        var ansr=solve(midx+1,r)%__MOD__;
        return (cur+ansl+ansr)%__MOD__;
    }

    return solve(0,arr.length-1);
};